arangod/Aql/Ast.cpp

```diff
@@ -2217,125 +2217,54 @@ std::unordered_set<std::string> Ast::getReferencedAttributesForKeep(AstNode cons
   return result;
 }
 
-bool Ast::populateSingleAttributeAccess(AstNode const* node,
-                                        Variable const* variable,
-                                        std::vector<std::string>& attributeName) {
-  bool result = true;
-
-  attributeName.clear();
-  std::vector<std::string> attributePath;
-
-  auto visitor = [&](AstNode const* node) {
-    if (!result) {
-      return false;
-    }
-
-    if (node->type == NODE_TYPE_ATTRIBUTE_ACCESS) {
-      attributePath.emplace_back(node->getString());
-      return true;
-    }
-
-    if (node->type == NODE_TYPE_REFERENCE) {
-      // reference to a variable
-      auto v = static_cast<Variable const*>(node->getData());
-
-      if (v == nullptr) {
-        THROW_ARANGO_EXCEPTION(TRI_ERROR_OUT_OF_MEMORY);
-      }
-
-      if (v->id == variable->id) {
-        // the variable we are looking for
-        if (attributeName.empty()) {
-          // haven't seen an attribute before. so store the attribute we got
-          attributeName = std::move(attributePath);
-        } else {
-          // have seen some attribute before. now check if it's the same attribute
-          size_t const n = attributeName.size();
-          if (n != attributePath.size()) {
-            // different attributes
-            result = false;
-          } else {
-            for (size_t i = 0; i < n; ++i) {
-              if (attributePath[i] != attributeName[i]) {
-                // different attributes
-                result = false;
-                break;
-              }
-            }
-          }
-        }
-      }
-      // fall-through
-    }
-
-    attributePath.clear();
-    return true;
-  };
-
-  traverseReadOnly(node, visitor, doNothingVisitor);
-  if (attributeName.empty()) {
-    return false;
-  }
-
-  return result;
-}
-
-/// @brief checks if the only references to the specified variable are
-/// attribute accesses to the specified attribute. all other variables
-/// used in the expression are ignored and will not influence the result!
-bool Ast::variableOnlyUsedForSingleAttributeAccess(AstNode const* node,
-                                                   Variable const* variable,
-                                                   std::vector<std::string> const& attributeName) {
-  bool result = true;
-
+/// @brief determines the top-level attributes referenced in an expression for the
+/// specified out variable
+bool Ast::getReferencedAttributes(AstNode const* node, 
+                                  Variable const* variable,
+                                  std::unordered_set<std::string>& vars) {
```

Comment:
Is there a reason why this cannot be an attributePath?
e.g.:
`a.b.c` would not be optimized t was before